src/main/java/com/google/devtools/build/lib/analysis/Runfiles.java

```diff
@@ -1155,35 +1070,25 @@ public Runfiles mergeAll(Sequence<?> sequence, StarlarkThread thread) throws Eva
     }
   }
 
-  /**
-   * Fingerprint this {@link Runfiles} tree.
-   */
-  public void fingerprint(Fingerprint fp) {
+  /** Fingerprint this {@link Runfiles} tree. */
+  public void fingerprint(ActionKeyContext actionKeyContext, Fingerprint fp) {
     fp.addBoolean(legacyExternalRunfiles);
     fp.addPath(suffix);
-    Map<PathFragment, Artifact> symlinks = getSymlinksAsMap(null);
-    fp.addInt(symlinks.size());
-    for (Map.Entry<PathFragment, Artifact> symlink : symlinks.entrySet()) {
-      fp.addPath(symlink.getKey());
-      fp.addPath(symlink.getValue().getExecPath());
-    }
-    Map<PathFragment, Artifact> rootSymlinks = getRootSymlinksAsMap(null);
-    fp.addInt(rootSymlinks.size());
-    for (Map.Entry<PathFragment, Artifact> rootSymlink : rootSymlinks.entrySet()) {
-      fp.addPath(rootSymlink.getKey());
-      fp.addPath(rootSymlink.getValue().getExecPath());
-    }
-
-    for (Artifact artifact : artifacts.toList()) {
-      fp.addPath(artifact.getRunfilesPath());
-      fp.addPath(artifact.getExecPath());
-    }
-
-    for (String name : getEmptyFilenames().toList()) {
-      fp.addString(name);
-    }
+    try {
+      actionKeyContext.addNestedSetToFingerprint(SYMLINK_ENTRY, fp, symlinks);
+      actionKeyContext.addNestedSetToFingerprint(SYMLINK_ENTRY, fp, rootSymlinks);
+      actionKeyContext.addNestedSetToFingerprint(RUNFILES_AND_EXEC_PATH, fp, artifacts);
+      actionKeyContext.addNestedSetToFingerprint(fp, getEmptyFilenames());
+    } catch (CommandLineExpansionException | InterruptedException e) {
+      // This should never happen as the MapFn implementations used above don't throw these
+      // exceptions.
+      throw new IllegalStateException(e);
+    }
+
+    // Not fingerprinting the names of empty files as they are a pure function of the other entries.
   }
-  /** Describes the inputs {@link fingerprint} uses to aid describeKey() descriptions. */
+
+  /** Describes the inputs {@link #fingerprint} uses to aid describeKey() descriptions. */
   public String describeFingerprint() {
```

Comment:
Can you update the describeFingerprint output to note that the reported results may be slightly misleading because addNestedSetToFingerprint will fingerprint the *exact* underlying structure of the nested sets, not simply the logical elements in the encountered order? It's easy to look at the output and see that the values are "equal", but you can't see the underlying nested set structure that the fingerprinting uses.

This tripped me up for a day trying to diagnose an action conflict. All the inputs were logically equivalent, but some difference in the NestedSet structures caused the fingerprinting to be different.

Reply:
Was this comment meant to apply to my other PR?

Reply:
I meant in this PR, though I'd give the same comment in your other PR (18419, right?).

In this PR, the fingerprint() function (which is called by computeKey somewhere, right?) is being changed to call addNestedSetToFingerprint(), which will include the exact internal structure of the nested set in the fingerprint.

Meanwhile, describeFingerprint will just render the flattened, linearized, unique, elements of the nested set. The net effect is, if an action conflict occurs, you get an error like:

```
Action conflict:
computeKey fingerprint: 8Ef7283Ag, F83E241
files: a, b, c
symlinks: d, e, f
rootSymlinks: g, h, i
```

i.e. the contents of the two are shown to be identical. Which is confusing, because if they're the same, why is there a problem? But it can be hard to identify that they *are* the same if the output is large.

There's another case of these elsewhere, I think something to do with tree artifacts; describeFingerprint can't accurately/correctly tell what is in them. Which is fine, we just added a line to the output to mention the reported info may not be entirely accurate. Then you have a hint that when the fingerprint differs, but everything else looks the same, it might be something to do with describeFingerprint not being able to be 100% accurate, so you can dig deeper into that object to try and identify the difference another way.